test-data/unit/check-narrowing.test

```diff
@@ -2206,3 +2206,45 @@ def f3(x: object) -> None:
     else:
         reveal_type(x)  # N: Revealed type is "builtins.object"
 [builtins fixtures/primitives.pyi]
+
+[case testConsistentNarrowingEqAndIn]
+# flags: --python-version 3.10
+
+# https://github.com/python/mypy/issues/17864
+def f(x: str | int) -> None:
+    if x == "x":
+        reveal_type(x)  # N: Revealed type is "Union[builtins.str, builtins.int]"
```

Comment:
Is this desired behavior? I feel like we should narrow to `Literal["x"]`. Yes, it might be a subclass of int that compares equal with str, but don't do that.

Reply:
Yeah, this case seems fine. But classes that compare equal to `int` are not unusual (e.g. `Decimal`), so it's not clear where to draw the line. We might need some ad-hoc rules to support common use cases.

Reply:
The last time I tried that change it looked pretty good on primer, I think the main obstacle was it exposed some shortcomings in the binder